base_model_restrict_update/models/ir_model_access.py

```diff
@@ -17,19 +17,37 @@ def check(self, model, mode="read", raise_exception=True):
         if self.env.su:
             return True
         res = super().check(model, mode, raise_exception)
-        self._cr.execute(
-            "SELECT restrict_update FROM ir_model WHERE model = %s", (model,)
-        )
-        query_res = self._cr.dictfetchone()
-        if (
-            query_res["restrict_update"]
-            and mode != "read"
-            and not self.env.user.unrestrict_model_update
-        ):
-            if raise_exception:
+        if mode != "read" and raise_exception:
+            if self._test_readonly(model) or self._test_restrict_update(model):
                 raise AccessError(
                     _("You are only allowed to read this record. (%s - %s)")
                     % (model, mode)
                 )
-            return False
         return res
+
+    @api.model
+    def _test_readonly(self, model):
+        exclude_models = self._readonly_exclude_models()
+        if model not in exclude_models and self.env.user.is_readonly_user:
+            return True
+        return False
+
+    @api.model
+    def _test_restrict_update(self, model):
+        self.env.cr.execute(
+            "SELECT restrict_update FROM ir_model WHERE model = %s", (model,)
+        )
+        query_res = self.env.cr.dictfetchone()
+        if query_res["restrict_update"] and not self.env.user.unrestrict_model_update:
+            return True
+        return False
+
+    @api.model
+    def _readonly_exclude_models(self):
+        """ Models updtate/create by system, and should be excluded from checking """
```

Comment:
I'm not sure, why we need this?

Reply:
Because res.user itself will be updated when switch back and forth, does it make sense?

Reply:
:thinking:  could you describe a use case where this is needed?